apps/rush-lib/src/logic/buildCache/ProjectBuildCache.ts

```diff
@@ -0,0 +1,220 @@
+    this._terminal = options.terminal;
+  }
+
+  public async tryHydrateFromCacheAsync(): Promise<boolean> {
+    const cacheId: string | undefined = this._cacheId;
+    if (!cacheId) {
+      this._terminal.writeWarningLine('Unable to get cache ID. Ensure Git is installed.');
+      return false;
+    }
+
+    const cacheEntryBuffer:
+      | Buffer
+      | undefined = await this._buildCacheProvider.tryGetCacheEntryBufferByIdAsync(this._terminal, cacheId);
+    if (!cacheEntryBuffer) {
+      this._terminal.writeVerboseLine('No build cache hit.');
+      return false;
+    }
+
+    this._terminal.writeLine('Build cache hit.');
+
+    const projectFolderPath: string = this._project.projectFolder;
+
+    // Purge output folders
+    this._terminal.writeVerboseLine(`Clearing cached folders: ${this._projectOutputFolderNames.join(', ')}`);
+    await Promise.all(
+      this._projectOutputFolderNames.map((outputFolderName: string) =>
+        FileSystem.deleteFolderAsync(path.join(projectFolderPath, outputFolderName))
+      )
+    );
+
+    const tarStream: stream.Writable = tar.extract({ cwd: projectFolderPath });
+    const success: boolean = await new Promise(
+      (resolve: (result: boolean) => void, reject: (error: Error) => void) => {
+        try {
+          tarStream.on('error', (error: Error) => reject(error));
+          tarStream.on('close', () => resolve(true));
+          tarStream.on('drain', () => resolve(true));
+          tarStream.write(cacheEntryBuffer);
+        } catch (e) {
+          reject(e);
+        }
+      }
+    );
+
+    if (success) {
+      this._terminal.writeLine('Successfully hydrated build output from cache.');
+    } else {
+      this._terminal.writeWarningLine('Hydration build output from cache was unsuccessful.');
+    }
+
+    return success;
+  }
+
+  public async trySetCacheEntryAsync(): Promise<boolean> {
+    const cacheId: string | undefined = this._cacheId;
+    if (!cacheId) {
+      this._terminal.writeWarningLine('Unable to get cache ID. Ensure Git is installed.');
+      return false;
+    }
+
+    const projectFolderPath: string = this._project.projectFolder;
+    const outputFoldersThatExist: boolean[] = await Promise.all(
+      this._projectOutputFolderNames.map((outputFolderName) =>
+        FileSystem.existsAsync(path.join(projectFolderPath, outputFolderName))
+      )
+    );
+    const filteredOutputFolders: string[] = [];
+    for (let i: number = 0; i < outputFoldersThatExist.length; i++) {
+      if (outputFoldersThatExist[i]) {
+        filteredOutputFolders.push(this._projectOutputFolderNames[i]);
+      }
+    }
+
+    this._terminal.writeVerboseLine(
+      `Caching existent build output folders: ${filteredOutputFolders.join(', ')}`
+    );
+    const tarStream: stream.Readable = tar.create(
+      {
+        gzip: true,
+        portable: true,
+        cwd: projectFolderPath
+      },
+      filteredOutputFolders
```

Comment:
What happens if am output folder contains symlinks?  Should we report an error?

It might be better to fail during caching than to fail during restoring, or to restore an unusable output.